lib/src/block_parser.dart

```diff
@@ -728,7 +730,137 @@ class ParagraphSyntax extends BlockSyntax {
       parser.advance();
     }
 
-    var contents = parser.document.parseInline(childLines.join('\n'));
-    return new Element('p', contents);
+    var paragraphLines = _extractReflinkDefinitions(parser, childLines);
+    if (paragraphLines == null) {
+      // Paragraph consisted solely of reference link definitions.
+      return new Text('');
+    } else {
+      var contents = new UnparsedContent(paragraphLines.join('\n'));
+      return new Element('p', [contents]);
+    }
+  }
+
+  // Extract reference link definitions from the front of the paragraph, and
+  // return the remaining paragraph lines.
+  List<String> _extractReflinkDefinitions(
+      BlockParser parser, List<String> lines) {
```

Comment:
This function _is_ hairy 😟  It took me a few re-writes to even get this 😛 .

However, I think your suggestion won't work, for one annoying reason: Link reference definitions, which look like:

> [label]: http://url "title of many words"

are allowed to not have a title, and they're allowed to have a title on the line following the URL, and the title is allowed to be multiline, so the following are all link reference definitions:

> [label]: http://url
>
> [label]: http://url
> "title"
>
> [label]: http://url
> "title
> with many
> words"

So (I think), we have to do this nutty thing that I'm doing, first sectioning the "paragraph" into multiple possible link reference definitions, then looking closely at each one to see a) if it even is a real link reference definition, and b) what's the "longest" link reference definition to be found.

So I don't know how to parse certain examples like [157](http://spec.commonmark.org/0.25/#example-157) and [159](http://spec.commonmark.org/0.25/#example-159).